clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java

```diff
@@ -233,15 +240,120 @@ public void testPollEvent() {
     }
 
     @Test
-    public void testSubscriptionChangeEvent() {
-        SubscriptionChangeEvent event = new SubscriptionChangeEvent();
+    public void testTopicSubscriptionChangeEvent() {
+        Set<String> topics = Set.of("topic1", "topic2");
+        Optional<ConsumerRebalanceListener> listener = Optional.of(new MockRebalanceListener());
+        TopicSubscriptionChangeEvent event = new TopicSubscriptionChangeEvent(topics, listener, 12345);
 
         setupProcessor(true);
+        when(subscriptionState.subscribe(topics, listener)).thenReturn(true);
+        when(metadata.requestUpdateForNewTopics()).thenReturn(1);
         when(heartbeatRequestManager.membershipManager()).thenReturn(membershipManager);
         processor.process(event);
+
+        verify(subscriptionState).subscribe(topics, listener);
+        verify(metadata).requestUpdateForNewTopics();
+        assertEquals(1, processor.metadataVersionSnapshot());
+        verify(membershipManager).onSubscriptionUpdated();
+        // verify member state doesn't transition to JOINING.
+        verify(membershipManager, never()).onConsumerPoll();
+        assertDoesNotThrow(() -> event.future().get());
+    }
+
+    @Test
+    public void testTopicSubscriptionChangeEventWithIllegalSubscriptionState() {
+        subscriptionState = new SubscriptionState(new LogContext(), OffsetResetStrategy.EARLIEST);
+        Optional<ConsumerRebalanceListener> listener = Optional.of(new MockRebalanceListener());
+        TopicSubscriptionChangeEvent event = new TopicSubscriptionChangeEvent(
+            Set.of("topic1", "topic2"), listener, 12345);
+
+        subscriptionState.subscribe(Pattern.compile("topic.*"), listener);
+        setupProcessor(true);
+        when(metadata.requestUpdateForNewTopics()).thenReturn(1);
+        when(heartbeatRequestManager.membershipManager()).thenReturn(membershipManager);
+        processor.process(event);
+
+        ExecutionException e = assertThrows(ExecutionException.class, () -> event.future().get());
+        assertInstanceOf(IllegalStateException.class, e.getCause());
```

Comment:
should we add an assertion on the e.getMessage() too? would make sure we got the state exception we want, and it also makes the test intention really explicit.